Documentation for debug rendering component (#1860)
o3de/o3de.org: content/docs/user-guide/components/reference/atom/debug-rendering.md

---
title: Debug Render Component
linktitle: Debug Render Component

> description: 'Open 3D Engine (O3DE) debug rendering level component reference.'

Reviewer: ```suggestion
description: 'Use the Debug Rendering level component with Atom Renderer in Open 3D Engine (O3DE) '
```

{{< image-width "/images/user-guide/components/reference/atom/debug-rendering/debug-rendering-component.jpg" "500" "Debug Render Component" >}}

## Debug Visualization Properties

The **Debug Rendering Level Component** is used to visualize rendering information about the scene, such as material properties like albedo and roughness or lighting factors like direct/indirect and diffuse/specular.

| Property | Description |
|-|-|
| **Enable Render Debugging** | Whether to utilize render debugging or not. If set to false, properties in this component have no effect. |
| **Debug View Mode** | Specifies what debug information is displayed in the viewport (for example material albedo, metalness, roughness...). Note that the options "Albedo" and "BaseColor" differ in that BaseColor will display the color used for metal reflectance, whereas Albedo for metals will be zero/black. |
| **Lighting Type** | Whether to display Diffuse Lighting, Specular Lighting, or both. |
| **Lighting Source** | Whether to display Direct Lighting, Indirect Lighting, both, or use a debug directional light as the only light source (see the next section). |

> Screeshot with **Debug View Mode = Albedo**

Reviewer: Move to **Examples** section at the bottom of the page (per earlier comment on reorganizing).
```suggestion
Example with **Debug View Mode** set to `Albedo`:
```

{{< image-width "/images/user-guide/components/reference/atom/debug-rendering/albedo.jpg" "1500" "Screeshot with Debug View Mode set to Albedo" >}}

> Screenshot with **Lighting Type = Diffuse** and **Lighting Source = Indirect** (i.e. the scene will only apply indirect diffuse lighting)

Reviewer: Move to **Examples** section at the bottom of the page (per earlier comment on reorganizing).
```suggestion
Example of debug lighting, with **Lighting Type** set to `Diffuse` and **Lighting Source** set to `Indirect`:
```

| **Override Roughness** | Whether to override the roughness for all materials in the scene to the value specified below. |
| **Roughness Value** | The value used to override roughness on all materials in the scene. For example you can set this to zero if you want to visualize a very glossy and reflective scene. |
| **Override Metallic** | Whether to override the metallic for all materials in the scene to the value specified below. |
| **Metallic Value** | The value used to override metallic on all materials in the scene. For example you can set this one to make all materials in your scene metals. |
| **Enable Normal Maps** | Whether to enable/disable normal maps on all materials in the scene (note that disabling normal maps will also disable detail normal maps, and thus only the vertex normal will be used in shading calculations). |
| **Enable Detail Normal Maps** | Whether to enable/disable detail normal maps on all materials in the scene (note that not all materials provide or support detail normal maps, therefore those materials will be unaffected by this option). |

> Screenshot overriding **Base Color** to **Grey (R:128, G:128, B:128)**

Reviewer: Move to **Examples** section at the bottom of the page (per earlier comment on reorganizing).
```suggestion
Example of debug materials, with **Override Base Color** set to grey (R: `128`, G: `128`, B: `128`):
```

| **Override Roughness** | Whether to override the roughness for all materials in the scene to the value specified below. | Off |
| **Roughness Value** | The value used to override roughness on all materials in the scene. For example you can set this to zero if you want to visualize a very glossy and reflective scene. | 1.0 |
| **Override Metallic** | Whether to override the metallic for all materials in the scene to the value specified below. | Off |
| **Metallic Value** | The value used to override metallic on all materials in the scene. For example you can set this one to make all materials in your scene metals. | 0.0 |
| **Enable Normal Maps** | Whether to enable/disable normal maps on all materials in the scene (note that disabling normal maps will also disable detail normal maps, and thus only the vertex normal will be used in shading calculations). | On |
| **Enable Detail Normal Maps** | Whether to enable/disable detail normal maps on all materials in the scene (note that not all materials provide or support detail normal maps, therefore those materials will be unaffected by this option). | On |

## Custom Debug Properties

> When writing shaders, it can be useful to have immediate access to a tweakable value for debug purposes. The Debug Render Component provides access to **four debug booleans** and **four debug floats** that are passed into the Scene Shader Resource Group. These values are accessible from any shader that includes the Scene SRG.

Reviewer: ```suggestion
When debugging shaders, it can be useful to have immediate access to tweakable values. The Debug Render component provides access to four _debug booleans_ and four _debug floats_. These values are passed into the Scene Shader Resource Group (SRG) and are accessible from any shader that includes `SceneSRG`. For more information on SRGs, see [Shader Resource Groups](/docs/atom-guide/dev-guide/shaders/azsl/#shader-resource-groups).
```

> 

Reviewer: Move this here, and add links to the code. 

```suggestion
The custom debug float and boolean properties are exposed to shaders via the [SceneSrg.azsli](https://github.com/o3de/o3de/blob/development/Gems/Atom/Feature/Common/Assets/ShaderResourceGroups/SceneSrg.azsli) and [Debug.azsli](https://github.com/o3de/o3de/blob/development/Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Debug.azsli) files.
```

> **NOTE**: It is imperative that you delete any usage of these variables from your shaders when submitting your code. These values are strictly to facilitate local debugging and aren't meant for any other use. If you fail to delete your usage of these variable before you submit, another shader author might experience unwanted side effects when using these values.

Reviewer: ```suggestion
{{< important >}}
When submitting your shader code, you must delete any usage of these variables from your shaders. These values are strictly used to facilitate local debugging and aren't meant for any other use. If you don't delete these variable in your shader code before you submit, another shader author may experience unwanted side effects when using these values.
{{< /important >}}
```

Reviewer 2: This callout is directed at contributors assuming they are writing shaders for the O3DE project. Seems a bit out of place here as this topic is about using this debug rendering component, not shader authoring. Should this be moved elsewhere in the docs, perhaps to the SRG docs above? If not, can it be rephrased or clarified for developers who are working on game projects? "When contributing your shader code", what if they are working locally and not contributing to any project? Does this advice apply to runtime shaders in general?

Author: These aren't variable you would normally use while authoring shaders, they would only be used in tandem with the debug component, so this is a good place. As for game projects, this advice still applies, it is best never to check in usage of these variables as that may then conflict with someone else's usage (similar to how you wouldn't use printf debugging and then check that in after you've finished debugging). It's purely there for convenience so the user doesn't have to do the plumbing themselves if they want to modify shader variables at runtime.

Reviewer: Given the context above, I've revised this to the following: 
> These variables are only meant to help you debug your shaders locally. For best practice, when you are done debugging, we recommend that you delete any usage of these variables. It's especially important to delete them if you are contributing shader code either to the source repo or your team's own repo. If you don't, another shader author may experience unwanted side effects when using the same variables.

> The custom debug float and boolean properties are exposed to shaders via the SceneSrg.azsli and Debug.azsli files.

Reviewer: Move this above the **note**
```suggestion
```
